File: web/src/api/business/apiKey.js
import service from '@/utils/request'

// @Tags APIKey
// @Summary 创建API密钥
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.APIKey true "创建API密钥"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /api-key [post]
export const createAPIKey = (data) => {
  return service({
    url: '/api-key',
    method: 'post',
    data
  })
}

// @Tags APIKey
// @Summary 获取API密钥列表
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /api-key/list [get]
export const getAPIKeys = () => {
  return service({
    url: '/api-key/list',
    method: 'get'
  })
}

// @Tags APIKey
// @Summary 吊销API密钥
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.APIKey true "吊销API密钥"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /api-key/revoke [post]
export const revokeAPIKey = (data) => {
  return service({
    url: '/api-key/revoke',
    method: 'post',
    data
  })
}

// @Tags APIKey
// @Summary 删除API密钥
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body dbModel.APIKey true "删除API密钥"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /api-key [delete]
export const deleteAPIKey = (data) => {
  return service({
    url: '/api-key',
    method: 'delete',
    data
  })
}

// @Tags APIKey
// @Summary 验证API密钥
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body object{key string} true "验证API密钥"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /api-key/validate [post]
export const validateAPIKey = (data) => {
  return service({
    url: '/api-key/validate',
    method: 'post',
    data
  })
}

File: server/api/v1/business/api_key.go
package business

import (
	"time"

	"github.com/flipped-aurora/gin-vue-admin/server/model/business"
	"github.com/flipped-aurora/gin-vue-admin/server/model/common/request"
	"github.com/flipped-aurora/gin-vue-admin/server/model/common/response"
	"github.com/flipped-aurora/gin-vue-admin/server/utils"
	"github.com/gin-gonic/gin"
)

type APIKeyApi struct{}

func (b *APIKeyApi) Create(c *gin.Context) {
	var req struct {
		Name      string `json:"name" binding:"required"`
		ExpiresAt int64  `json:"expires_at"`
	}
	err := c.ShouldBindJSON(&req)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}

	userID := utils.GetUserID(c)
	var expiresAt *time.Time
	if req.ExpiresAt > 0 {
		t := time.Unix(req.ExpiresAt, 0)
		expiresAt = &t
	}

	apiKey, rawKey, err := APIKeyServiceApp.GenerateAPIKey(req.Name, userID, expiresAt)
	if err != nil {
		response.FailWithMessage("创建失败: "+err.Error(), c)
		return
	}

	response.OkWithDetailed(gin.H{
		"api_key": apiKey,
		"raw_key": rawKey,
	}, "创建成功，请妥善保存 raw_key", c)
}

func (b *APIKeyApi) GetAPIKeys(c *gin.Context) {
	apiKeys, err := APIKeyServiceApp.GetAPIKeys()
	if err != nil {
		response.FailWithMessage("获取失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(apiKeys, "获取成功", c)
}

func (b *APIKeyApi) Revoke(c *gin.Context) {
	var reqId request.GetById
	err := c.ShouldBindJSON(&reqId)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = APIKeyServiceApp.RevokeAPIKey(reqId.Uint())
	if err != nil {
		response.FailWithMessage("撤销失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("撤销成功", c)
}

func (b *APIKeyApi) Delete(c *gin.Context) {
	var reqId request.GetById
	err := c.ShouldBindJSON(&reqId)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = APIKeyServiceApp.DeleteAPIKey(reqId.Uint())
	if err != nil {
		response.FailWithMessage("删除失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("删除成功", c)
}

func (b *APIKeyApi) Validate(c *gin.Context) {
	key := c.GetHeader("X-API-Key")
	if key == "" {
		response.FailWithMessage("缺少 API Key", c)
		return
	}

	apiKey, err := APIKeyServiceApp.ValidateAPIKey(key)
	if err != nil || apiKey == nil {
		response.FailWithMessage("无效的 API Key", c)
		return
	}

	APIKeyServiceApp.UpdateLastUsed(apiKey.ID)

	response.OkWithDetailed(apiKey, "验证成功", c)
}
